### PR TITLE
Fixes #16759 - ensures OSM layers use OGR random layer read/write

### PR DESCRIPTION
**Pull request by stev-0**
Guarantee all features are rendered with large files.

## Description
Any OSM files larger than around 50MB (could be smaller) seemed to have trouble being rendered directly in QGIS via OGR. It seems that GDAL 2.2 introduced a new reading mode specifically designed for OSM files, which have a tendency to run out of memory if they are read sequentially due to the large number of nodes which must be stored in memory to create ways (linestrings). This was not being used in QGIS, so this PR just switches the new API on for OSM layers only.

More detail is available here https://trac.osgeo.org/gdal/wiki/rfc66_randomlayerreadwrite. Please see the issue on [redmine](https://issues.qgis.org/issues/16759) for pictures pre and post the PR.

Please could someone review the code here, my C++ is not yet very reliable.

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [x] Commit messages are descriptive and explain the rationale for changes
- [x] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [ ] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [ ] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and contain sufficient information in the commit message to be documented
- [x] I ave read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [ ] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit

### CODE REVIEW

**Comment by nyalldawson** [comment]
@rouault?

File: src/providers/ogr/qgsogrfeatureiterator.cpp [hunk 1]
```diff
@@ -282,20 +296,38 @@ bool QgsOgrFeatureIterator::fetchFeature( QgsFeature &feature )
 
   gdal::ogr_feature_unique_ptr fet;
 
-  while ( fet.reset( OGR_L_GetNextFeature( mOgrLayer ) ), fet )
+  // OSM layers (especially large ones) need the GDALDataset::GetNextFeature() call rather than OGRLayer::GetNextFeature()
+  // see more details here: https://trac.osgeo.org/gdal/wiki/rfc66_randomlayerreadwrite
+  if ( mSource->mDriverName == QLatin1String( "OSM" ) )
```
Comment (rouault) [inline src/providers/ogr/qgsogrfeatureiterator.cpp]: this should be protected by #ifdef GDAL_VERSION_NUM >= GDAL_COMPUTE_VERSION(2,2,0)

File: src/providers/ogr/qgsogrfeatureiterator.cpp [hunk 2]
```diff
@@ -282,20 +296,38 @@ bool QgsOgrFeatureIterator::fetchFeature( QgsFeature &feature )
...
-    if ( !mFilterRect.isNull() && ( !feature.hasGeometry() || feature.geometry().isEmpty() ) )
-      continue;
-
-    // we have a feature, end this cycle
-    feature.setValid( true );
-    geometryToDestinationCrs( feature, mTransform );
-    return true;
+    while ( fet.reset( GDALDatasetGetNextFeature( mConn->ds, nullptr, nullptr, nullptr, nullptr )), fet )
+    {
+      if ( checkFeature( fet, feature ) )
```
Comment (rouault) [inline src/providers/ogr/qgsogrfeatureiterator.cpp]: this should check that the feature belongs to the layer of interest. It is returned in the second parameter of GDALDatasetGetNextFeature()

File: src/providers/ogr/qgsogrfeatureiterator.cpp [hunk 3]
```diff
@@ -282,20 +296,38 @@ bool QgsOgrFeatureIterator::fetchFeature( QgsFeature &feature )
...
+  {
 
-  } // while
+    while ( fet.reset( OGR_L_GetNextFeature( mOgrLayer ) ), fet )
+    {
+      if ( checkFeature( fet, feature ) ) 
+      {
+        return true;
+      }
+      else
```
Comment (m-kuhn) [inline src/providers/ogr/qgsogrfeatureiterator.cpp]: Is the else branch here required?

File: src/providers/ogr/qgsogrfeatureiterator.cpp [hunk 4]
```diff
@@ -282,20 +296,41 @@ bool QgsOgrFeatureIterator::fetchFeature( QgsFeature &feature )
...
-    feature.setValid( true );
-    geometryToDestinationCrs( feature, mTransform );
-    return true;
+#if GDAL_VERSION_NUM >= GDAL_COMPUTE_VERSION(2,2,0)
+  if ( mSource->mDriverName == QLatin1String( "OSM" ) )
+  {
+    OGRLayerH nextFeatureBelongingLayer;
+    while ( fet.reset( GDALDatasetGetNextFeature( mConn->ds, &nextFeatureBelongingLayer, nullptr, nullptr, nullptr ) ), fet )
+    {
+      if ( checkFeature( fet, feature ) && QString::compare( QString::fromLatin1( OGR_L_GetName( nextFeatureBelongingLayer ) ), mSource->mLayerName ) == 0 )
```
Comment (nyalldawson) [inline src/providers/ogr/qgsogrfeatureiterator.cpp]: I think this check should be reversed -- the layer name comparison is much cheaper than the checkFeature, which requires reading the feature itself.

Comment (rouault) [reply src/providers/ogr/qgsogrfeatureiterator.cpp]: And you can replace the layer name comparison by checking the equality of layer pointers directly.

File: src/providers/ogr/qgsogrfeatureiterator.cpp [hunk 5]
```diff
@@ -305,7 +305,7 @@ bool QgsOgrFeatureIterator::fetchFeature( QgsFeature &feature )
     OGRLayerH nextFeatureBelongingLayer;
     while ( fet.reset( GDALDatasetGetNextFeature( mConn->ds, &nextFeatureBelongingLayer, nullptr, nullptr, nullptr ) ), fet )
     {
-      if ( checkFeature( fet, feature ) && QString::compare( QString::fromLatin1( OGR_L_GetName( nextFeatureBelongingLayer ) ), mSource->mLayerName ) == 0 )
+      if ( nextFeatureBelongingLayer == mSource->mLayerName && checkFeature( fet, feature ) )
```
Comment (rouault) [inline src/providers/ogr/qgsogrfeatureiterator.cpp]: huh ! does this actually compile ??? You're comparing a OGRLayer* pointer to a QString !

Comment (stev-0) [reply src/providers/ogr/qgsogrfeatureiterator.cpp]: My bad - not at my workstation, so shortcutting. That'll teach me

**Comment by stev-0** [comment]
This should be all done now. Let me know if I should squash commits

**Comment by nyalldawson** [comment]
Thanks @stev-0!

**Pull request merged**